### PR TITLE
Propagate (some) signals to subprocesses in json_run_localhost

### PR DESCRIPTION
**Pull request by dgquintas**
But more importantly, don't assert before interrupting and joining the subprocesses.

### CODE REVIEW

**Comment by ctiller** [comment]
Actually... let's change the vector to a fixed-size array and have an integer counter for the number of elements: it could be that we get a signal while creating the initial job list.

**Comment by dgquintas** [comment]
Done. PTAL.

**Comment by dgquintas** [comment]
@markdroth it seems we are not asserting anymore: https://github.com/grpc/grpc/commit/c56d3ebd7eadca14322230d97fdddedd9579f41a#diff-7c8f49d8a5a6ef474ee98afcc1894d50L94 is this intentional?

File: test/cpp/qps/json_run_localhost.cc [hunk 1]
```diff
@@ -42,6 +46,10 @@
 #include "test/cpp/util/subprocess.h"
 
 using grpc::SubProcess;
+typedef std::unique_ptr<SubProcess> SubProcessPtr;
+SubProcessPtr g_driver;
```
Comment (markdroth) [inline test/cpp/qps/json_run_localhost.cc]: The Google C++ style guide prohibits global variables of non-POD type:

https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

File: test/cpp/qps/json_run_localhost.cc [hunk 2]
```diff
@@ -42,6 +46,10 @@
...
+constexpr auto kNumWorkers = 2;
+std::vector<SubProcessPtr> g_workers(2);
```
Comment (markdroth) [inline test/cpp/qps/json_run_localhost.cc]: Same issue here.

Comment (ctiller) [inline test/cpp/qps/json_run_localhost.cc]: Recommend rewriting as:

static SubProcess *g_workers[2];

Comment (dgquintas) [inline test/cpp/qps/json_run_localhost.cc]: Done

File: test/cpp/qps/json_run_localhost.cc (hunk 1 above)
Comment (dgquintas) [inline test/cpp/qps/json_run_localhost.cc]: Changed it into a pointer. Done.

File: test/cpp/qps/json_run_localhost.cc [hunk 3]
```diff
@@ -61,8 +62,9 @@ std::string as_string(const T& val) {
 static void sighandler(int sig) {
   const int errno_saved = errno;
   g_driver->Interrupt();
```
Comment (markdroth) [inline test/cpp/qps/json_run_localhost.cc]: I think you need `if (g_driver != NULL)` here.

File: test/cpp/qps/json_run_localhost.cc [hunk 4]
```diff
@@ -112,20 +114,25 @@ int main(int argc, char** argv) {
...
+  for (int i = 0; i < kNumWorkers; ++i) {
+    if (g_workers[i]) {
+      const int worker_status = g_workers[i]->Join();
       if (worker_status != 0) {
         LogStatus(worker_status, "worker");
       }
     }
   }
+
+  delete g_driver;
```
Comment (markdroth) [inline test/cpp/qps/json_run_localhost.cc]: Do you want to do these two lines of deletion in the signal handler as well, just to be tidy?

File: test/cpp/qps/json_run_localhost.cc (hunk 3 above)
Comment (dgquintas) [inline test/cpp/qps/json_run_localhost.cc]: Done

File: test/cpp/qps/json_run_localhost.cc (hunk 4 above)
Comment (dgquintas) [inline test/cpp/qps/json_run_localhost.cc]: We can't do that. The signal can come at any time (in fact, if we ^C, it'll definitely come before the end of the main). If the handler deletes, the final lines of main that interrupt and join the workers and driver will incur in an use-after-free.

Comment (ctiller) [inline test/cpp/qps/json_run_localhost.cc]: More than that: you can't call malloc/free during signal handlers (or new/delete). To convince yourself of why: what happens if we get the signal during a free()? What are the guarantees about the allocators internal data structures?

**Pull request merged**